File: surface/widgets/compose/components/widgets/domain/MarkdownPreview.kt
// ============================================================
// Clef Surface Compose Widget — MarkdownPreview
//
// Live markdown rendering widget. Transforms raw markdown source
// into a Column of styled Text elements: headings as bold text,
// bold/italic via TextStyle, bullets for lists, indented styled
// text for blockquotes, and monospace text for code blocks.
//
// Adapts the markdown-preview.widget spec: anatomy (root,
// content), states (static, rendering), and connect attributes.
// ============================================================

package clef.surface.compose.components.widgets.domain

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Divider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// --------------- Types ---------------

private sealed class RenderedLine {
    data class Heading(val level: Int, val content: String) : RenderedLine()
    data class BoldLine(val content: String) : RenderedLine()
    data class TextLine(val content: String) : RenderedLine()
    data class Bullet(val content: String) : RenderedLine()
    data class Quote(val content: String) : RenderedLine()
    data class Code(val content: String) : RenderedLine()
    object Hr : RenderedLine()
    object Blank : RenderedLine()
}

// --------------- Helpers ---------------

private fun parseMarkdown(source: String): List<RenderedLine> {
    val lines = source.split("\n")
    val result = mutableListOf<RenderedLine>()
    var inCodeBlock = false

    for (line in lines) {
        if (line.startsWith("```")) {
            inCodeBlock = !inCodeBlock
            if (inCodeBlock) {
                result.add(RenderedLine.Code("--- code ---"))
            }
            continue
        }

        if (inCodeBlock) {
            result.add(RenderedLine.Code(line))
            continue
        }

        val trimmed = line.trim()

        if (trimmed.isEmpty()) {
            result.add(RenderedLine.Blank)
            continue
        }

        // Headings
        val headingMatch = Regex("^(#{1,6})\\s+(.*)").find(trimmed)
        if (headingMatch != null) {
            result.add(
                RenderedLine.Heading(
                    level = headingMatch.groupValues[1].length,
                    content = headingMatch.groupValues[2],
                ),
            )
            continue
        }

        // Horizontal rule
        if (Regex("^[-*_]{3,}$").matches(trimmed)) {
            result.add(RenderedLine.Hr)
            continue
        }

        // Blockquote
        if (trimmed.startsWith("> ")) {
            result.add(RenderedLine.Quote(trimmed.drop(2)))
            continue
        }

        // Bullet list
        if (Regex("^[-*+]\\s").containsMatchIn(trimmed)) {
            result.add(RenderedLine.Bullet(trimmed.drop(2)))
            continue
        }

        // Numbered list
        val numberedMatch = Regex("^\\d+\\.\\s(.*)").find(trimmed)
        if (numberedMatch != null) {
            result.add(RenderedLine.Bullet(numberedMatch.groupValues[1]))
            continue
        }

        // Bold text line
        if (trimmed.startsWith("**") && trimmed.endsWith("**")) {
            result.add(RenderedLine.BoldLine(trimmed.drop(2).dropLast(2)))
            continue
        }

        // Regular text
        result.add(RenderedLine.TextLine(trimmed))
    }

    return result
}

// --------------- Component ---------------

/**
 * Markdown preview rendered as a Column of styled Text elements.
 *
 * @param content Raw markdown content string.
 * @param modifier Modifier for the root layout.
 */
@Composable
fun MarkdownPreview(
    content: String,
    modifier: Modifier = Modifier,
) {
    val parsed = remember(content) { parseMarkdown(content) }

    Column(modifier = modifier.padding(8.dp)) {
        parsed.forEach { line ->
            when (line) {
                is RenderedLine.Heading -> {
                    val fontSize = when (line.level) {
                        1 -> 24.sp
                        2 -> 20.sp
                        3 -> 18.sp
                        else -> 16.sp
                    }
                    Text(
                        text = line.content,
                        style = TextStyle(
                            fontSize = fontSize,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.primary,
                        ),
                        modifier = Modifier.padding(vertical = 4.dp),
                    )
                }

                is RenderedLine.BoldLine -> {
                    Text(
                        text = line.content,
                        fontWeight = FontWeight.Bold,
                        modifier = Modifier.padding(vertical = 2.dp),
                    )
                }

                is RenderedLine.Bullet -> {
                    Row(modifier = Modifier.padding(start = 8.dp, top = 2.dp, bottom = 2.dp)) {
                        Text(
                            text = "\u2022",
                            color = MaterialTheme.colorScheme.primary,
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(text = line.content)
                    }
                }

                is RenderedLine.Quote -> {
                    Row(modifier = Modifier.padding(start = 16.dp, top = 2.dp, bottom = 2.dp)) {
                        Text(
                            text = "\u2502",
                            color = MaterialTheme.colorScheme.outlineVariant,
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(
                            text = line.content,
                            fontStyle = FontStyle.Italic,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }

                is RenderedLine.Code -> {
                    Text(
                        text = line.content,
                        fontFamily = FontFamily.Monospace,
                        color = Color(0xFFDCDCAA),
                        modifier = Modifier.padding(start = 16.dp, top = 1.dp, bottom = 1.dp),
                    )
                }

                is RenderedLine.Hr -> {
                    Divider(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp),
                        color = MaterialTheme.colorScheme.outlineVariant,
                    )
                }

                is RenderedLine.Blank -> {
                    Spacer(modifier = Modifier.height(8.dp))
                }

                is RenderedLine.TextLine -> {
                    Text(
                        text = line.content,
                        modifier = Modifier.padding(vertical = 2.dp),
                    )
                }
            }
        }
    }
}

File: surface/widgets/swiftui/components/widgets/domain/MarkdownPreview.swift
// ============================================================
// Clef Surface SwiftUI Widget — MarkdownPreview
//
// Live markdown rendering widget. Transforms raw markdown source
// into a VStack of styled Text elements: headings as bold text,
// bold/italic via font modifiers, bullets for lists, indented
// styled text for blockquotes, and monospace text for code blocks.
// ============================================================

import SwiftUI

private enum RenderedLine: Identifiable {
    case heading(level: Int, content: String)
    case boldLine(content: String)
    case textLine(content: String)
    case bullet(content: String)
    case quote(content: String)
    case code(content: String)
    case hr
    case blank

    var id: String {
        switch self {
        case .heading(let l, let c): return "h\(l)-\(c)"
        case .boldLine(let c): return "b-\(c)"
        case .textLine(let c): return "t-\(c)"
        case .bullet(let c): return "li-\(c)"
        case .quote(let c): return "q-\(c)"
        case .code(let c): return "c-\(c)"
        case .hr: return "hr-\(UUID().uuidString)"
        case .blank: return "blank-\(UUID().uuidString)"
        }
    }
}

private func parseMarkdown(_ source: String) -> [RenderedLine] {
    let lines = source.components(separatedBy: "\n")
    var result: [RenderedLine] = []
    var inCodeBlock = false

    for line in lines {
        if line.hasPrefix("```") {
            inCodeBlock.toggle()
            if inCodeBlock {
                result.append(.code("--- code ---"))
            }
            continue
        }

        if inCodeBlock {
            result.append(.code(line))
            continue
        }

        let trimmed = line.trimmingCharacters(in: .whitespaces)

        if trimmed.isEmpty {
            result.append(.blank)
            continue
        }

        // Headings
        if let match = trimmed.range(of: #"^(#{1,6})\s+(.*)"#, options: .regularExpression) {
            let hashPart = String(trimmed[trimmed.startIndex..<trimmed.firstIndex(of: " ")!])
            let content = String(trimmed[trimmed.index(after: trimmed.firstIndex(of: " ")!)...])
            result.append(.heading(level: hashPart.count, content: content))
            continue
        }

        // Horizontal rule
        if trimmed.range(of: #"^[-*_]{3,}$"#, options: .regularExpression) != nil {
            result.append(.hr)
            continue
        }

        // Blockquote
        if trimmed.hasPrefix("> ") {
            result.append(.quote(String(trimmed.dropFirst(2))))
            continue
        }

        // Bullet list
        if trimmed.range(of: #"^[-*+]\s"#, options: .regularExpression) != nil {
            result.append(.bullet(String(trimmed.dropFirst(2))))
            continue
        }

        // Numbered list
        if let match = trimmed.range(of: #"^\d+\.\s(.*)"#, options: .regularExpression) {
            let content = String(trimmed[trimmed.range(of: #"(?<=\d+\.\s).*"#, options: .regularExpression)!])
            result.append(.bullet(content))
            continue
        }

        // Bold text line
        if trimmed.hasPrefix("**") && trimmed.hasSuffix("**") && trimmed.count > 4 {
            result.append(.boldLine(String(trimmed.dropFirst(2).dropLast(2))))
            continue
        }

        // Regular text
        result.append(.textLine(trimmed))
    }

    return result
}

struct MarkdownPreviewView: View {
    var content: String

    var body: some View {
        let parsed = parseMarkdown(content)

        VStack(alignment: .leading, spacing: 2) {
            ForEach(Array(parsed.enumerated()), id: \.offset) { _, line in
                switch line {
                case .heading(let level, let text):
                    Text(text)
                        .font(level <= 1 ? .title : level == 2 ? .title2 : level == 3 ? .title3 : .headline)
                        .fontWeight(.bold)
                        .foregroundColor(.accentColor)
                        .padding(.vertical, 4)

                case .boldLine(let text):
                    Text(text)
                        .fontWeight(.bold)
                        .padding(.vertical, 2)

                case .bullet(let text):
                    HStack(alignment: .top, spacing: 8) {
                        Text("\u{2022}")
                            .foregroundColor(.accentColor)
                        Text(text)
                    }
                    .padding(.leading, 8)
                    .padding(.vertical, 2)

                case .quote(let text):
                    HStack(alignment: .top, spacing: 8) {
                        Text("\u{2502}")
                            .foregroundColor(Color(.systemGray3))
                        Text(text)
                            .italic()
                            .foregroundColor(.secondary)
                    }
                    .padding(.leading, 16)
                    .padding(.vertical, 2)

                case .code(let text):
                    Text(text)
                        .font(.system(size: 13, design: .monospaced))
                        .foregroundColor(Color(red: 0.86, green: 0.86, blue: 0.67))
                        .padding(.leading, 16)
                        .padding(.vertical, 1)

                case .hr:
                    Divider()
                        .padding(.vertical, 8)

                case .blank:
                    Spacer().frame(height: 8)

                case .textLine(let text):
                    Text(text)
                        .padding(.vertical, 2)
                }
            }
        }
        .padding(8)
    }
}
